packages/smooth_app/lib/background/background_task_manager.dart

```diff
@@ -122,11 +122,15 @@ class BackgroundTaskManager {
     }
   }
 
-  /// Runs a single task. Possible exception.
+  // TODO(monsieurtanuki): when we are more relaxed, enable the tasks to be tried several times.
+  /// Runs a single task, once and for all.
   Future<void> _runTask(final AbstractBackgroundTask task) async {
-    await task.execute(localDatabase);
-    await task.postExecute(localDatabase);
-    await _remove(task.uniqueId);
+    try {
+      await task.execute(localDatabase);
+      await task.postExecute(localDatabase);
+    } finally {
+      await _remove(task.uniqueId);
+    }
```

Comment:
Can't we get the eventual error to see if it's a network related error (in this case wait and retry) because we want to provide offline mode !

Reply:
Currently we can't:
* we would need a list of exceptions (why not?)
* the way it's coded now the task either succeeds or fails (with an exception)
* it was supposed to be a quick fix, and you said later that we didn't need it after all

There's a long story about background tasks, and it took me 1 month to refactor it in a PR and then 1 month to split my PR in 7 PRs because it was too complex to review. cf. #2968 and #3018.

Actually the most critical part was not in the background tasks themselves, but within the app: making sure that I could see the freshest product data from the server, with all my local changes on top, from any part of the app, and with immediate refresh.

Regarding the background tasks, I coded in a "I'm rather lucky" mode and assumed there were no tasks doomed to fail forever. Which is our main problem here: dealing with tasks that will fail forever.